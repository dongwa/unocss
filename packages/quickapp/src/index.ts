export * from './preset'
export * from './webpack'
