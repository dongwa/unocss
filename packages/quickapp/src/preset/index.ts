import type { Postprocessor, Preset, PresetOptions } from '@unocss/core'
// import { preflights } from './preflights'
import { defaultRules, transformEscapESelector } from 'unplugin-transform-class/utils'
import { rules } from './rules'
import type { Theme, ThemeAnimation } from './theme'
import { theme } from './theme'
import { variants } from './variants'
export { preflights } from './preflights'
export { theme, colors } from './theme'
export { parseColor } from './utils'
export { transformerQuickappAttributify, transformerClass, transformerDirectives } from './transformer'

export type { ThemeAnimation, Theme }

export interface DarkModeSelectors {
  /**
   * Selector for light variant.
   *
   * @default '.light'
   */
  light?: string

  /**
   * Selector for dark variant.
   *
   * @default '.dark'
   */
  dark?: string
}

export interface PresetQuickappOptions extends PresetOptions {
  /**
   * Dark mode options
   *
   * @default 'class'
   */
  dark?: 'class' | 'media' | DarkModeSelectors
  /**
   * Generate pesudo selector as `[group=""]` instead of `.group`
   *
   * @default false
   */
  attributifyPseudo?: Boolean
  /**
   * Prefix for CSS variables.
   *
   * @default 'un-'
   */
  variablePrefix?: string
  /**
   * Utils prefix
   *
   * @default undefined
   */
  prefix?: string
  /**
   * Generate preflight
   *
   * @default true
   */
  preflight?: boolean

  /**
   * 是否转换class
   *
   * @default true
   */
  transform?: boolean

  /**
   * 自定义转换规则
   * @default https://github.com/MellowCo/unplugin-transform-class#options
   */
  transformRules?: Record<string, string>
}

export const presetQuickapp = (options: PresetQuickappOptions = {}): Preset<Theme> => {
  options.dark = options.dark ?? 'class'
  options.attributifyPseudo = options.attributifyPseudo ?? false
  options.preflight = options.preflight ?? true

  options.transform = options.transform ?? true
  options.transformRules = options.transformRules ?? defaultRules

  return {
    name: '@unocss/preset-quickapp',
    theme: {
      ...theme,
      transformRules: options.transformRules,
    },
    rules,
    variants: variants(options),
    options,
    postprocess(css) {
      if (options.transform)
        css.selector = transformEscapESelector(css.selector, options.transformRules)
      if (options.variablePrefix && options.variablePrefix !== 'un-')
        VarPrefixPostprocessor(options.variablePrefix)
    },
    // preflights: options.preflight ? preflights : [],
    prefix: options.prefix,
  }
}

export default presetQuickapp

function VarPrefixPostprocessor(prefix: string): Postprocessor {
  return (obj) => {
    obj.entries.forEach((i) => {
      i[0] = i[0].replace(/^--un-/, `--${prefix}`)
      if (typeof i[1] === 'string')
        i[1] = i[1].replace(/var\(--un-/g, `var(--${prefix}`)
    })
  }
}
