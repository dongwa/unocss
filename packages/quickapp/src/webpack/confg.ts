import type { UserConfig } from '@unocss/core'
import remToPxPreset from '@unocss/preset-rem-to-px'
import { presetQuickapp, transformerDirectives } from '../preset/index'

export const defaultOutput = 'src/css/uno.css'
export const defaultInclude = [/\.(ux|[jt]sx|html)($|\?)/]
export const defaultRules: Record<string, string> = {
  '.': '-d-',
  '/': '-s-',
  ':': '-c-',
  '%': '-p-',
  '!': '-e-',
  '#': '-w-',
  '(': '-bl-',
  ')': '-br-',
  '[': '-fl-',
  ']': '-fr-',
  '$': '-r-',
}

export interface UnocssQuickappOptions<Theme extends {} = {}> extends UserConfig<Theme> {
  /** 配置自动转义class的规则 */
  transformRules: Record<string, string>
  /** 配置unocss输出路径 */
  unoCssOutput: string
}

export function resolveConfig<Theme extends {}>(config: Partial<UnocssQuickappOptions<Theme>> = {}) {
  const defaultConfig: UnocssQuickappOptions<Theme> = {
    include: defaultInclude,
    unoCssOutput: defaultOutput,
    transformRules: defaultRules,
    transformers: [transformerDirectives(
      {
        enforce: 'pre',
      },
    )],
    presets: [remToPxPreset() as any, presetQuickapp(
      {
        transformRules: defaultRules,
      },
    )],
  }
  config = {
    ...defaultConfig,
    ...config,
  }
  return config as UnocssQuickappOptions<Theme>
}

