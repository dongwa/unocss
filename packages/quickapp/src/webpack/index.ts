import path from 'path'
import { existsSync, readFileSync, writeFileSync } from 'fs'
import type { UserConfig, UserConfigDefaults } from '@unocss/core'
import type { ResolvedUnpluginOptions, UnpluginOptions } from 'unplugin'

import { createUnplugin } from 'unplugin'
import WebpackSources from 'webpack-sources'
import toolkitStyle from '@aiot-toolkit/compiler/lib/style'
import remToPxPreset from '@unocss/preset-rem-to-px'
import { presetQuickapp } from '../preset/index'
import { createContext } from '../../../shared-integration/src/context'
import { getPath, isCssId } from '../../../shared-integration/src/utils'
import { applyTransformers } from '../../../shared-integration/src/transformers'
import { HASH_PLACEHOLDER_RE, LAYER_MARK_ALL, RESOLVED_ID_RE, getHashPlaceholder, getLayerPlaceholder, resolveId, resolveLayer } from '../../../shared-integration/src/layers'

export interface WebpackPluginOptions<Theme extends {} = {}> extends UserConfig<Theme> {
  /** 配置自动转义class的规则 */
  transformRules?: Record<string, string>
  /** 配置unocss输出路径 */
  unoCssOutput?: string
}

const PLUGIN_NAME = 'unocss:quickapp'
export const LAYER_PLACEHOLDER_RE = /(\")#--unocss--\":\s*{\s*\"layer\":\s*\"(.+?)?\"\s*}/g
export const defaultInclude = [/\.(ux|[jt]sx|html)($|\?)/]
let OUTPUTCSS = 'src/css/uno.css'
let defaultRules: Record<string, string> = {
  '.': '-d-',
  '/': '-s-',
  ':': '-c-',
  '%': '-p-',
  '!': '-e-',
  '#': '-w-',
  '(': '-bl-',
  ')': '-br-',
  '[': '-fl-',
  ']': '-fr-',
  '$': '-r-',
}

export function defineConfig<Theme extends {}>(config: WebpackPluginOptions<Theme>) {
  return config
}

export function UnoQuickappWebpackPlugin<Theme extends {}>(
  config?: WebpackPluginOptions<Theme>,
) {
  return createUnplugin(() => {
    /** 初始化配置,默认内置所需的presets */
    OUTPUTCSS = config?.unoCssOutput || OUTPUTCSS
    defaultRules = config?.transformRules || defaultRules

    const defaultConfig: UserConfigDefaults & { include?: string } = {
      presets: [remToPxPreset(), presetQuickapp(
        {
          transformRules: defaultRules,
        },
      )],
    }

    const ctx = createContext<WebpackPluginOptions>({
      include: defaultInclude,
      ...config as any,
    }, defaultConfig)
    const { uno, tokens, filter, extract/** , onInvalidate */ } = ctx
    const nonPreTransformers = ctx.uno.config.transformers?.filter(i => i.enforce !== 'pre')
    if (nonPreTransformers?.length) {
      console.warn(
        // eslint-disable-next-line prefer-template
        '[unocss] webpack integration only supports "pre" enforce transformers currently.'
        + 'the following transformers will be ignored\n'
        + nonPreTransformers.map(i => ` - ${i.name}`).join('\n'),
      )
    }

    const tasks: Promise<any>[] = []
    const entries = new Set<string>()
    const hashes = new Map<string, string>()

    const plugin = <UnpluginOptions>{
      name: PLUGIN_NAME,
      enforce: 'pre',
      transformInclude(id) {
        return filter('', id) && !id.match(/\.html$/) && !RESOLVED_ID_RE.test(id)
      },
      async transform(code, id) {
        const result = await applyTransformers(ctx, code, id, 'pre')
        if (isCssId(id))
          return result
        if (result == null)
          tasks.push(extract(code, id))
        else
          tasks.push(extract(result.code, id))
        return result
      },
      buildStart() {
        /** 开始时创建uno.css文件 */
        const layer = getLayerPlaceholder(LAYER_MARK_ALL)
        if (!existsSync(OUTPUTCSS)) { writeFileSync(OUTPUTCSS, layer) }
        else {
          let currentUnoCss = readFileSync(OUTPUTCSS).toString()
          if (!currentUnoCss.includes(layer)) {
            currentUnoCss = `${layer}\n${currentUnoCss}`
            writeFileSync(OUTPUTCSS, currentUnoCss)
          }
        }
      },
      resolveId(id) {
        const entry = resolveId(id)
        if (entry === id)
          return
        if (entry) {
          let query = ''
          const queryIndex = id.indexOf('?')
          if (queryIndex >= 0)
            query = id.slice(queryIndex)
          entries.add(entry)
          // preserve the input query
          return entry + query
        }
      },
      loadInclude(id) {
        const layer = getLayer(id)
        return !!layer
      },
      // serve the placeholders in virtual module
      load(id) {
        const layer = getLayer(id)
        const hash = hashes.get(id)
        if (layer)
          return (hash ? getHashPlaceholder(hash) : '') + getLayerPlaceholder(layer)
      },
      webpack(compiler) {
        // replace the placeholders
        compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation) => {
          compilation.hooks.optimizeAssets.tapPromise(PLUGIN_NAME, async () => {
            const files = Object.keys(compilation.assets)

            await Promise.all(tasks)
            const result = await uno.generate(tokens, { minify: true })

            for (const file of files) {
              // https://github.com/unocss/unocss/pull/1428
              if (file === '*')
                return

              let code = compilation.assets[file].source().toString()
              let replaced = false
              code = code.replace(HASH_PLACEHOLDER_RE, '')
              code = code.replace(LAYER_PLACEHOLDER_RE, (_, quote, layer) => {
                replaced = true
                const css = layer === LAYER_MARK_ALL
                  ? result.getLayers(undefined, Array.from(entries)
                    .map(i => resolveLayer(i)).filter((i): i is string => !!i))
                  : result.getLayer(layer) || ''
                const filePath = path.resolve(ctx.root, 'build', file)
                // eslint-disable-next-line no-console
                console.log('css', css)
                const res = toolkitStyle.parse({ code: css, filePath })
                // eslint-disable-next-line no-console
                console.log('result', res.jsonStyle)
                return JSON.stringify(res.jsonStyle).slice(1, -1)
              })
              /** 转化代码中不支持的转义class */
              code = transformCode(code)
              if (replaced)
                compilation.assets[file] = new WebpackSources.RawSource(code) as any
            }
          })
        })
      },
    } as Required<ResolvedUnpluginOptions>

    return plugin
  }).webpack()
}

function getLayer(id: string) {
  let layer = resolveLayer(getPath(id))
  if (!layer) {
    const entry = resolveId(id)
    if (entry)
      layer = resolveLayer(entry)
  }
  return layer
}

/**
 * @desc 去左右空格
 * @param value - 需要处理的字符串
 */
export function trim(value: string) {
  return value.replace(/(^\s*)|(\s*$)/g, '')
}

/**
 * 获取class
 */
export function getClass(code: string) {
  const matchs: string[][] = []
  // ux
  Array.from(code.matchAll(/\"classList\"\s*:\s*\[((?:\n|.)*?)\]/g)).forEach((m) => {
    const classStr = m[1]
    const sourceStr = trim(m[0])

    const classArr = [sourceStr]
    classArr.push(classStr)
    matchs.push(classArr)
  })
  return matchs
}

function escapeRegExp(str = '') {
  return str.replace(/[/.*+?^${}()|[\]\\]/g, '\\$&')
}

function createTransformRegExp(rules: Record<string, string>) {
  return new RegExp(`[${escapeRegExp(Object.keys(rules).join(''))}]`)
}

export function transformCode(code: string, rules = defaultRules) {
  const classNames = getClass(code)

  classNames.forEach((c) => {
    let currentClass = c[0]
    c.slice(1).forEach((selector) => {
      currentClass = currentClass.replace(selector, transformEscapESelector(selector, rules))
    })
    code = code.replace(c[0], currentClass)
  })

  return code
}

export function transformEscapESelector(selector = '', rules = defaultRules): string {
  const transformRegExp = createTransformRegExp(rules)
  if (transformRegExp.test(selector)) {
    for (const transformRule in rules) {
      const replaceReg = new RegExp(escapeRegExp(`${transformRule}`), 'g')
      selector = selector.replace(replaceReg, rules[transformRule])
    }
  }
  return selector
}

