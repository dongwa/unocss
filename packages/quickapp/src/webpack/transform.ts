import { defaultRules } from './confg'

/**
 * @desc 去左右空格
 * @param value - 需要处理的字符串
 */
export function trim(value: string) {
  return value.replace(/(^\s*)|(\s*$)/g, '')
}

/**
   * 获取class
   */
export function getClass(code: string) {
  const matchs: string[][] = []
  // ux
  Array.from(code.matchAll(/\"classList\"\s*:\s*(?:function\s*\(\)\s*\{return)?\s*\[((?:\n|.)*?)\]/g)).forEach((m) => {
    const classStr = m[1]
    const sourceStr = trim(m[0])

    let classArr = [sourceStr]
    classArr = classArr.concat(getArrClass(classStr))
    // classArr.push(classStr)
    matchs.push(classArr)
  })
  return matchs
}

export function getArrClass(className: string) {
  // [
  //   title === '2.3' ? 'font-$font-name bg-teal-200:55' :'tracking-[2/5]',
  //   isFont ? 'font-$font-name' : 'tracking-[2/5]'
  // ]
  // => ['font-$font-name bg-teal-200:55', 'tracking-[2/5]','font-$font-name', 'tracking-[2/5]']
  return Array.from(className.matchAll(/[\"\'](.*?)[\"\']/g)).map(v => v[1])
}

function escapeRegExp(str = '') {
  return str.replace(/[/.*+?^${}()|[\]\\]/g, '\\$&')
}

function createTransformRegExp(rules: Record<string, string>) {
  return new RegExp(`[${escapeRegExp(Object.keys(rules).join(''))}]`)
}

export function transformCode(code: string, rules = defaultRules) {
  const classNames = getClass(code)

  classNames.forEach((c) => {
    let currentClass = c[0]
    c.slice(1).forEach((selector) => {
      currentClass = currentClass.replace(selector, transformEscapESelector(selector, rules))
    })
    code = code.replace(c[0], currentClass)
  })

  return code
}

export function transformEscapESelector(selector = '', rules = defaultRules): string {
  const transformRegExp = createTransformRegExp(rules)
  if (transformRegExp.test(selector)) {
    for (const transformRule in rules) {
      const replaceReg = new RegExp(escapeRegExp(`${transformRule}`), 'g')
      selector = selector.replace(replaceReg, rules[transformRule])
    }
  }
  return selector
}

