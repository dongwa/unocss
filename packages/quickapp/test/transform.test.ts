import { describe, expect, test } from 'vitest'
import { getClass, transformCode, transformEscapESelector } from '../src/webpack'

describe('transformer', () => {
  test('transformEscapESelector', () => {
    const test1 = '\n    "w-1.4",\n    "flex:col",\n    "w-1/3",\n    "90%",\n    "flex!row"\n  '
    const result = '\n    "w-1-d-4",\n    "flex-c-col",\n    "w-1-s-3",\n    "90-p-",\n    "flex-e-row"\n  '
    const res = transformEscapESelector(test1)
    // eslint-disable-next-line no-console
    console.log('res', res)

    expect(res).toBe(result)
  })

  test('funcClass', () => {
    // "classList": [
    //   "flex",
    //   "flex-col",
    //   "justify-around",
    //   "items-center",
    //   "w-full",
    //   "p30"
    // ],
    const test2 = '"classList": function () {return [\'w-20\', \'h-20\', \'border-1\', \'border-gray-300\', \'text-center\', \'content-center\', \'focus:bg-gray-500\', this.getClass()]}'
    // const result = '"classList": function () {return [\'w-20\', \'h-20\', \'border-1\', \'border-gray-300\', \'text-center\', \'content-center\', \'focus-c-bg-gray-500\', this.getClass()]}'
    const classStr = getClass(test2)

    expect(classStr).toMatchInlineSnapshot(`
      [
        [
          "\\"classList\\": function () {return ['w-20', 'h-20', 'border-1', 'border-gray-300', 'text-center', 'content-center', 'focus:bg-gray-500', this.getClass()]",
          "w-20",
          "h-20",
          "border-1",
          "border-gray-300",
          "text-center",
          "content-center",
          "focus:bg-gray-500",
        ],
      ]
    `)
    const res = transformCode(test2.toString())
    expect(res).toMatchInlineSnapshot('"\\"classList\\": function () {return [\'w-20\', \'h-20\', \'border-1\', \'border-gray-300\', \'text-center\', \'content-center\', \'focus-c-bg-gray-500\', this.getClass()]}"')
  })
})
