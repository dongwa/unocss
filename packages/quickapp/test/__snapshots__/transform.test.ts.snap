// Vitest Snapshot v1

exports[`transformer > funcClass 1`] = `[]`;
